### DISCUSSION TITLE
How to configure exception log color background style for failed tests

### DISCUSSION THREAD

**Discussion by chrisyarbrough** [cake-build/cake#4069] [Q&A] [answered]
I'm running unit tests and would like to stop the build process if they fail. This seems to work fine by throwing an exception, but it results in the error text being renderer with a red background and a very low contrast font color in my IDE and Terminal:

<img width="746" alt="image" src="https://user-images.githubusercontent.com/17833862/201490516-b0b468ae-fb0d-4797-a7a7-162c9944dc52.png">

I saw that there's a feature request for configuring the color palette, but in my case I wonder if there's a way to simply disable the background color styling for this type of message. I'd rather have no backgrounds, but red text instead.

If there's no builtin feature yet, is this something I can overwrite by creating my own module? I'm not sure which class to replace though. I tried looking through the source code and stumbled over ICakeLog, IConsole and IConsoleRenderer, but I'm not seeing where the red background color is actually set caused by the exception, so I wouldn't know where to insert my changes.

Also, is this the usual way to report test failures? It feels a little weird to throw an exception and then read "An error occurred" in the log rather than something like "Tests failed, build aborted". I tried to find an option to set the task status (Succeeded, Failed, Skipped), but didn't see it. Maybe it would be a better idea to add a Criterion to follow-up build tasks to skip them when the test fails, but not throw an exception, and instead run to the end to handle test reporting as part of a custom reporter?

Sorry for being a Cake noob and thanks for any answers! I love the work being done here; would be happy to invest some time as well, when I get to a point where I understand Cake. :)

**Comment by nils-a** [comment, marked as answer]
For Reference: Color-related issues are #1461, #2976, #3138, #3245 

@chrisyarbrough It seems that is not possible, currently. However, simply ignoring all colors should be fairly easy to do in a [module](https://cakebuild.net/docs/extending/modules/):

The `IConsole` has the ability to change the color - if you supply a version of `IConsole` that simply "ignores" changes to `ForegroundColor` and `BackgroundColor` you'll get output completely without color changes.

## Alternative solution with custom colors

While the `IConsoleRenderer` decides which color to use to render logs, supplying a version of `IConsoleRenderer` would not change much, since the `ICakeLog` decides on its own which `IConsoleRenderer` to use:

https://github.com/cake-build/cake/blob/9828d7b246d332054896e52ba56983822feb3f05/src/Cake.Core/Diagnostics/CakeBuildLog.cs#L19-L41

So, if you want to define your own colors, you'd need to supply a custom version of `IConsoleRenderer`, as well as a custom version of `ICakeLog` that uses this renderer.

## As to the other question:

> Also, is this the usual way to report test failures? 

Normally, I'd use a unit test project and have that run using some console runner for the given unit test framework of my choosing. Then again, I've never written things for Unitiy. So, I wouldn't know. Sorry.
I've found [this post on medium](https://medium.com/geekculture/doing-basic-unit-testing-in-unity-c-9de7f923d2ee) which suggest using NUnit. In that case, I would rather use the [NUnit commandline runner](https://cakebuild.net/dsl/nunit-v3/). But, as I said: I've never done Unity before.
